### PR TITLE
Use rubygems to build tar without changing directory

### PR DESCRIPTION
**Pull request by tlunter**
Alternative to #257

### CODE REVIEW

**Comment by tlunter** [comment]
@nahiluhmot @bfulton This implementation uses rubygem's tar_writer instead of archive-tar-minitar

File: lib/docker/util.rb [hunk 1]
```diff
@@ -123,14 +123,34 @@ def create_tar(hash = {})
...
-  ensure
-    FileUtils.cd(cwd)
+  end
+
+  def fixed_file_names(directory)
+    Find.find(directory) do |file|
+      stat = File.stat(file)
+      next unless stat.file?
+
+      directory = directory + '/' unless directory.end_with?('/')
```
Comment (bfulton) [inline lib/docker/util.rb]: Could go above the `Find.find` and use `+=`.

File: lib/docker/util.rb [hunk 2]
```diff
@@ -123,14 +123,34 @@ def create_tar(hash = {})
...
+
+      file_name = file.split(directory)[1]
```
Comment (bfulton) [inline lib/docker/util.rb]: Something like `file_name = file[directory.length..-1]` is simpler and safer — `split` would do the wrong thing when the directory is present again later in the path.

File: lib/docker/util.rb [hunk 3]
```diff
@@ -123,14 +123,34 @@ def create_tar(hash = {})
...
+        end
+      end
+    end
+
     File.new(tempfile.path, 'r')
-  ensure
-    FileUtils.cd(cwd)
+  end
+
+  def fixed_file_names(directory)
```
Comment (bfulton) [inline lib/docker/util.rb]: This method might be better named something like `abs_and_relative_paths`.

File: lib/docker/util.rb [hunk 4]
```diff
@@ -123,14 +123,34 @@ def create_tar(hash = {})
...
-    tempfile_name = Dir::Tmpname.create('out') {}
-    tempfile = File.open(tempfile_name, 'wb+')
-    FileUtils.cd(directory)
-    Archive::Tar::Minitar.pack('.', tempfile)
+    tempfile = create_temp_file
+    Gem::Package::TarWriter.new(tempfile) do |tar|
+      fixed_file_names(directory) do |prefixed_file_name, file_name|
+        stat = File.stat(prefixed_file_name)
+        tar.add_file_simple(file_name, stat.mode, stat.size) do |tar_file|
+          tar_file.write(File.read(prefixed_file_name, mode: 'rb'))
```
Comment (bfulton) [inline lib/docker/util.rb]: Consider using `IO.copy_stream` to avoid reading the file into memory.

File: lib/docker/util.rb (hunk 3 above)
Comment (bfulton) [inline lib/docker/util.rb]: On second though, I don't think breaking this method out is helping, so I would inline it. You're calling `File.stat` twice per file, for example.

File: lib/docker.rb [hunk 5]
```diff
@@ -3,8 +3,8 @@
 require 'excon'
 require 'tempfile'
 require 'base64'
+require 'find'
```
Comment (bfulton) [inline lib/docker.rb]: Don't we need a `require 'rubygems/package'` somewhere?

**Comment by bfulton** [comment]
@tlunter this is a great approach: fixing the race by getting rid of the contention on cwd. Net win on dependencies also. A few comments above.

File: lib/docker.rb (hunk 5 above)
Comment (tlunter) [inline lib/docker.rb]: It's the line right below this comment :D

**Comment by tlunter** [comment]
@bfulton updated

File: lib/docker.rb (hunk 5 above)
Comment (bfulton) [inline lib/docker.rb]: Heh, no wonder it works!

**Comment by bfulton** [comment]
Looks good to me!

**Comment by Tranquility** [comment]
It looks like this changes the mtime and therefore breaks the caching of ADD in Dockerfiles.

**Comment by tlunter** [comment]
@Tranquility that is true :\ Looks like the rubygems code we're using doesn't support passing the current mtime in.  We might have to pull that code in to get that functionality since I haven't found a tar library for ruby that I've liked besides this.

**Pull request merged**